examples/tracing.cpp

```diff
@@ -0,0 +1,153 @@
+#include <memory>
+
+#include <jsapi.h>
+
+#include "boilerplate.h"
+
+// This example illustrates how to safely store GC pointers in the embedding's
+// data structures by implementing appropriate tracing mechanisms.
+
+////////////////////////////////////////////////////////////
+
+// An example C++ type that stores an arbitrary JS value.
+struct SafeBox {
+  // Arbitrarily JS value that will need to be traced. The JS::Heap type has a
+  // constructor and destructor and will trigger any required GC barriers on
+  // read or write.
+  JS::Heap<JS::Value> stashed;
+
+  // If we provide a trace method, we can then be used in a JS::Rooted or
+  // similar and the GC will be able to successfuly trace us.
+  void trace(JSTracer* trc) { JS::TraceEdge(trc, &stashed, "stashed value"); }
+};
+
+static bool CustomTypeExample(JSContext* cx) {
+  // If we use SafeBox as a stack object, then a JS::Rooted is enough.
+  JS::Rooted<SafeBox> stackSafe(cx);
+
+  // We can also use js::UniquePtr if SafeBox should be allocated on heap.
+  JS::Rooted<js::UniquePtr<SafeBox>> heapSafe(cx, js::MakeUnique<SafeBox>());
+
+  // NOTE: A JS::Rooted<SafeBox*> is a compile error. If one wanted to support
+  // rooting bare non-GC pointers then both JS::MapTypeToRootKind and
+  // JS::GCPolicy need to be defined for SafeBox*. This should be avoided in
+  // favor of using a smart-pointer when possible.
+
+  return true;
+}
+
+////////////////////////////////////////////////////////////
+
+// By defining a JS::GCPolicy we can support tracing in existing types that we
+// cannot add a trace method to such as std::shared_ptr.
+//
+// In this example, we forward methods to GCPolicy of target type and provide
+// reasonable behaviours when we have no current target.
+template <typename T>
+struct JS::GCPolicy<std::shared_ptr<T>> {
+  static void trace(JSTracer* trc, std::shared_ptr<T>* tp, const char* name) {
+    if (T* target = tp->get()) {
+      GCPolicy<T>::trace(trc, target, name);
+    }
+  }
+  static bool needsSweep(std::shared_ptr<T>* tp) {
+    if (T* target = tp->get()) {
+      return GCPolicy<T>::needsSweep(target);
+    }
+    return false;
+  }
+  static bool isValid(const std::shared_ptr<T>& t) {
```

Comment:
...and it looks like our docs don't include this isValid() requirement. Bug 1534440.